### DISCUSSION TITLE
Make implicitly typed variables non-nullable unless you are assigning a nullable value to them.

### DISCUSSION THREAD

**Discussion by ekolis** [dotnet/csharplang#4317] [Language Ideas] [answered]
I have the nullable feature enabled.

If I declare a string literal:

```
var name = "Fred";
```

then that will create my variable as a `string?` even though `var name = null;` is invalid.

And if I declare a variable to take the return value of a method or property that is not nullable, I still get a nullable variable:

```
var files = Directory.GetFiles(rootPath, $"*.{extension}", SearchOption.AllDirectories);
```

`files` is created as `string[]?` instead of `string[]` which is the actual return type of `Directory.GetFiles`.

This "hoisting" variables into nullable types is really annoying and makes the `var` keyword almost useless when the nullable feature is enabled because it almost never does what is intended, which is create a variable of the same type as the right hand side of the equals sign.

If this is necessary for "legacy reasons", can we at least get an alternate keyword for declaring an implicitly typed non-nullable variable? Maybe `var!` or `let` or something like that?

Thanks! 🙂

**Comment by YairHalberstadt** [comment, marked as answer]
This was an explicit decision - see https://github.com/dotnet/csharplang/blob/master/meetings/2019/LDM-2019-12-18.md#var, https://github.com/dotnet/csharplang/issues/3591, https://github.com/dotnet/csharplang/issues/3662

**Comment by HaloFour** [comment]
The feature was originally designed and shipped as you described and it was found to cause too many false warnings on reassignment, so the team reversed course and decided to have all inferred variables be nullable and rely on flow analysis to warn if the variable was then used in a manner that could throw.

**Reply by ekolis** [reply]
Oh, that's interesting. I just now got a chance to play around with it, must have missed that initial release. Any chance of that mode of operation becoming an optional flag in the project file, same as the nullable feature itself?

**Reply by 333fred** [reply]
No, we are not going to be adding a dialect for this.

**Reply by HaloFour** [reply]
Sounds like something that could be solved with an analyzer.  Detect that a `var` initially assigned with a non-null reference is later assigned to a nullable reference and emit a warning.
